<div style="border:solid blue 2px; padding: 20px">
  
**Hello Laura.**

My name is Yulia, and I will be reviewing your project. 

You will find my comments in coloured cells marked as 'Reviewer's comment'. The cell colour will vary based on the contents - I am explaining it further below. 

**Note:** Please do not remove or change my comments - they will help me in my future reviews and will make the process smoother for both of us. 

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment</b> 
    
Such comment will mark efficient solutions and good ideas that can be used in other projects.
</div>

<div class="alert alert-warning"; style="border-left: 7px solid gold">
<b>⚠️ Reviewer's comment</b> 
    
The parts marked with yellow comments indicate that there is room for optimisation. Though the correction is not necessary it is good if you implement it.
</div>

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment</b> 
    
If you see such a comment, it means that there is a problem that needs to be fixed. Please note that I won't be able to accept your project until the issue is resolved.
</div>

You are also very welcome to leave your comments / describe the corrections you've done / ask me questions, marking them with a different colour. You can use the example below: 

<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>

<div style="border:solid green 2px; padding: 20px">
    
<div class="alert alert-success">
<b>Review summary</b> 
    
Laura, thanks for submitting the project. You've done an outstanding job and I enjoyed reviewing it.
    
- You completed all the tasks.
- Your code was optimal and easy to read. 
- You wrote your own function and a loop very efficiently.
    
I've got no critical comments. It was a great start. 
    
The project is complete now. Good luck to you with your further studies and next projects.

## Basic Python - Project <a id='intro'></a>

## Introduction <a id='intro'></a>
In this project, you will work with data from the entertainment industry. You will study a dataset with records on movies and shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages 
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. There is no information about the quality of the data, so you will need to explore it before doing the analysis.

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.
 
Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Title and introduction are essential parts of the project. Make sure you do not forget to include it in your further projects. 
    
It is optimal if introduction consists of:
    
- brief description of the situation;
- goal of the project;
- description of the data we are going to use.
</div>


## Stage 1. Data overview <a id='data_review'></a>

Open and explore the data.

You'll need `pandas`, so import it.

In [1]:
# importing pandas
import pandas as pd


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
It's a good practice to upload all necessary libraries in the first code cell, separately from datasets (so that you do not have to reload the dataset if you need to add another library). 

Read the `movies_and_shows.csv` file from the `datasets` folder and save it in the `df` variable:

In [2]:
# reading the files and storing them to df
df = pd.read_csv('/datasets/movies_and_shows.csv')


Print the first 10 table rows:

In [3]:
# obtaining the first 10 rows from the df table
# hint: you can use head() and tail() in Jupyter Notebook without wrapping them into print()
df.head(10)


name                      Character   r0le        TITLE   Type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

Obtain the general information about the table with one command:

In [4]:
# obtaining general information about the data in df
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Great - you've used a comprehensive set of methods to have a first look at the data.

The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'. 


### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, we need to preprocess the data.

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Please note that it is highly recommended to add a conclusion / summary after each section and describe briefly your observations and / or major outcomes of the analysis.

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers and deal with the missing values. Then, check whether there are duplicates in the data.

In [5]:
# the list of column names in the df table
df.columns


Index(['   name', 'Character', 'r0le', 'TITLE', '  Type', 'release Year',
       'genres', 'imdb sc0re', 'imdb v0tes'],
      dtype='object')

Change the column names according to the rules of good style:
* If the name has several words, use snake_case
* All characters must be lowercase
* Remove whitespace
* Replace zero with letter 'o'

In [6]:
# renaming columns
df.rename(columns = {'   name':'name', 'Character':'character', 'r0le':'role', 'TITLE':'title', '  Type':'type', 'release Year':'release_year', 'imdb sc0re':'imdb_score','imdb v0tes':'imdb_votes'}, inplace = True)


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
This is a good way to rename the columns. And you set good names, in line with Python naming convention.

Check the result. Print the names of the columns once more:

In [7]:
# checking result: the list of column names
df.columns


Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imdb_score', 'imdb_votes'],
      dtype='object')

### Missing values <a id='missing_values'></a>
First, find the number of missing values in the table. To do so, combine two `pandas` methods:

In [8]:
# calculating missing values
df.isna().sum()


name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64

Not all missing values affect the research: the single missing value in `'title'` is not critical. The missing values in columns `'imdb_score'` and `'imdb_votes'` represent around 6% of all records (4,609 and 4,726, respectively, of the total 85,579). This could potentially affect our research. To avoid this issue, we will drop rows with missing values in the `'imdb_score'` and `'imdb_votes'` columns.

In [9]:
# dropping rows where columns with title, scores and votes have missing values
df = df.dropna(axis='rows')
df.head()


name                character   role        title   type  \
0   Robert De Niro            Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster            Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                      Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                    Betsy  ACTOR  Taxi Driver  MOVIE   

   release_year              genres  imdb_score  imdb_votes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0

Make sure the table doesn't contain any more missing values. Count the missing values again.

In [10]:
# counting missing values
df.isna().sum()


name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Well done, you've dropped the missing values successfully.

### Duplicates <a id='duplicates'></a>
Find the number of duplicate rows in the table using one command:

In [11]:
# counting duplicate rows
df.duplicated().sum()


6994

Review the duplicate rows to determine if removing them would distort our dataset.

In [12]:
# Produce table with duplicates (with original rows included) and review last 5 rows
duplicates = df[df.duplicated(keep=False)==True]
duplicates.head()


name                  character      role  \
7560      Philip Greene  Baseball Fan (uncredited)     ACTOR   
7561      Philip Greene  Baseball Fan (uncredited)     ACTOR   
14502          Dan Levy                   Reporter     ACTOR   
14512          Dan Levy                   Reporter     ACTOR   
18951  Nicolas Le Nev??                    unknown  DIRECTOR   

                                 title   type  release_year  \
7560                   How Do You Know  MOVIE          2010   
7561                   How Do You Know  MOVIE          2010   
14502  A Very Harold & Kumar Christmas  MOVIE          2011   
14512  A Very Harold & Kumar Christmas  MOVIE          2011   
18951                       Sammy & Co   SHOW          2014   

                                 genres  imdb_score  imdb_votes  
7560     ['comedy', 'drama', 'romance']         5.4     50383.0  
7561     ['comedy', 'drama', 'romance']         5.4     50383.0  
14502  ['comedy', 'fantasy', 'romance']         6.2     69562.0  
14512  ['comedy', 'fantasy', 'romance']         6.2     69562.0  
18951         ['animation', 'european']         5.7        31.0

There are two clear duplicates in the printed rows. We can safely remove them.
Call the `pandas` method for getting rid of duplicate rows:

In [13]:
# removing duplicate rows
df = df.drop_duplicates().reset_index(drop=True)
df.columns

Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imdb_score', 'imdb_votes'],
      dtype='object')

Check for duplicate rows once more to make sure you have removed all of them:

In [14]:
# checking for duplicates
df.duplicated().sum()


0

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Great, you found and removed the duplicates.

Now get rid of implicit duplicates in the `'type'` column. For example, the string `'SHOW'` can be written in different ways. These kinds of errors will also affect the result.

Print a list of unique `'type'` names, sorted in alphabetical order. To do so:
* Retrieve the intended dataframe column 
* Apply a sorting method to it
* For the sorted column, call the method that will return all unique column values

In [15]:
# viewing unique type names
df['type'].unique()


array(['MOVIE', 'the movie', 'tv show', 'shows', 'movies', 'tv shows',
       'tv series', 'tv', 'SHOW'], dtype=object)

Look through the list to find implicit duplicates of `'show'` (`'movie'` duplicates will be ignored since the assumption is about shows). These could be names written incorrectly or alternative names of the same genre.

You will see the following implicit duplicates:
* `'shows'`
* `'SHOW'`
* `'tv show'`
* `'tv shows'`
* `'tv series'`
* `'tv'`

To get rid of them, declare the function `replace_wrong_show()` with two parameters: 
* `wrong_shows_list=` — the list of duplicates
* `correct_show=` — the string with the correct value

The function should correct the names in the `'type'` column from the `df` table (i.e., replace each value from the `wrong_shows_list` list with the value in `correct_show`).

In [16]:
# function for replacing implicit duplicates
def replace_wrong_show(wrong_shows_list,correct_show):
    for wrong_show in wrong_shows_list:
        df['type']=df['type'].replace(wrong_show,correct_show)
    return df


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Very good. Please also note that you can go without the loop and pass the list directly to `replace()` function: 
    
    def replace_wrong_show(wrong_shows_list,correct_show):
    
        df['type']=df['type'].replace(wrong_shows_list,correct_show)
        return df


Call `replace_wrong_show()` and pass it arguments so that it clears implicit duplicates and replaces them with `SHOW`:

In [17]:
# removing implicit duplicates
replace_wrong_show(['tv show','tv shows','tv series','tv','shows'],'SHOW')


name                character      role        title  \
0          Robert De Niro            Travis Bickle     ACTOR  Taxi Driver   
1            Jodie Foster            Iris Steensma     ACTOR  Taxi Driver   
2           Albert Brooks                      Tom     ACTOR  Taxi Driver   
3           Harvey Keitel  Matthew 'Sport' Higgins     ACTOR  Taxi Driver   
4         Cybill Shepherd                    Betsy     ACTOR  Taxi Driver   
...                   ...                      ...       ...          ...   
73854       A??da Morales                  Maritza     ACTOR      Lokillo   
73855    Adelaida Buscato               Mar??a Paz     ACTOR      Lokillo   
73856  Luz Stella Luengas             Karen Bayona     ACTOR      Lokillo   
73857        In??s Prieto                    Fanny     ACTOR      Lokillo   
73858      Julian Gaviria                  unknown  DIRECTOR      Lokillo   

            type  release_year              genres  imdb_score  imdb_votes  
0          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
1          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
2          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
3          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
4          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
...          ...           ...                 ...         ...         ...  
73854  the movie          2021          ['comedy']         3.8        68.0  
73855  the movie          2021          ['comedy']         3.8        68.0  
73856  the movie          2021          ['comedy']         3.8        68.0  
73857  the movie          2021          ['comedy']         3.8        68.0  
73858  the movie          2021          ['comedy']         3.8        68.0  

[73859 rows x 9 columns]

Make sure the duplicate names are removed. Print the list of unique values from the `'type'` column:

In [18]:
# viewing unique genre names
df['type'].unique()


array(['MOVIE', 'the movie', 'SHOW', 'movies'], dtype=object)

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Confirmed, no implicit duplicates left.

### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed. 

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to our analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous project stages, you can now define how the assumption will be checked. Calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

Based on this, complete the following steps:

- Filter the dataframe to only include shows released in 1999 or later.
- Group scores into buckets by rounding the values of the appropriate column (a set of 1-10 integers will help us make the outcome of our calculations more evident without damaging the quality of our research).
- Identify outliers among scores based on their number of votes, and exclude scores with few votes.
- Calculate the average votes for each score and check whether the assumption matches the results.

To filter the dataframe and only include shows released in 1999 or later, you will take two steps. First, keep only titles published in 1999 or later in our dataframe. Then, filter the table to only contain shows (movies will be removed).

In [19]:
# using conditional indexing modify df so it has only titles released after 1999 (with 1999 included)
# give the slice of dataframe new name
goldenshows=df[df['release_year']>=1999]
goldenshows.tail()


name     character      role    title       type  \
73854       A??da Morales       Maritza     ACTOR  Lokillo  the movie   
73855    Adelaida Buscato    Mar??a Paz     ACTOR  Lokillo  the movie   
73856  Luz Stella Luengas  Karen Bayona     ACTOR  Lokillo  the movie   
73857        In??s Prieto         Fanny     ACTOR  Lokillo  the movie   
73858      Julian Gaviria       unknown  DIRECTOR  Lokillo  the movie   

       release_year      genres  imdb_score  imdb_votes  
73854          2021  ['comedy']         3.8        68.0  
73855          2021  ['comedy']         3.8        68.0  
73856          2021  ['comedy']         3.8        68.0  
73857          2021  ['comedy']         3.8        68.0  
73858          2021  ['comedy']         3.8        68.0

In [20]:
# repeat conditional indexing so df has only shows (movies are removed as result)
goldenshows=goldenshows[goldenshows['type']=='SHOW']

goldenshows.head()


name                character   role      title  type  \
1661       Jeff Probst           Himself - Host  ACTOR   Survivor  SHOW   
2073     Mayumi Tanaka  Monkey D. Luffy (voice)  ACTOR  One Piece  SHOW   
2074      Kazuya Nakai     Roronoa Zoro (voice)  ACTOR  One Piece  SHOW   
2075     Akemi Okamura             Nami (voice)  ACTOR  One Piece  SHOW   
2076  Kappei Yamaguchi            Usopp (voice)  ACTOR  One Piece  SHOW   

      release_year                                             genres  \
1661          2000                                        ['reality']   
2073          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2074          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2075          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2076          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   

      imdb_score  imdb_votes  
1661         7.4     24687.0  
2073         8.8    117129.0  
2074         8.8    117129.0  
2075         8.8    117129.0  
2076         8.8    117129.0

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
You've filtered the data correctly.

The scores that are to be grouped should be rounded. For instance, titles with scores like 7.8, 8.1, and 8.3 will all be placed in the same bucket with a score of 8.

In [21]:
# rounding column with scores
goldenshows['imdb_score'].tail()
goldenshows['imdb_score']=goldenshows['imdb_score'].round(0)

#checking the outcome with tail()
goldenshows['imdb_score'].tail()


73831    8.0
73832    8.0
73833    8.0
73834    8.0
73835    8.0
Name: imdb_score, dtype: float64

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Good job on rounding the score.

It is now time to identify outliers based on the number of votes.

In [22]:
# Use groupby() for scores and count all unique values in each group, print the result
scores_group = goldenshows.groupby('imdb_score')['imdb_votes'].count()
print(scores_group)


imdb_score
2.0       24
3.0       27
4.0      180
5.0      592
6.0     2494
7.0     4706
8.0     4842
9.0      557
10.0       8
Name: imdb_votes, dtype: int64


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
You've calculated the number of each value correctly.

Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores (we identified a range of 4-9 as acceptable), use conditional filtering and grouping.

In [23]:
# filter dataframe using two conditions (scores to be in the range 4-9)
goldenshows=goldenshows[(goldenshows['imdb_score']>=4)&(goldenshows['imdb_score']<=9)]

# group scores and corresponding average number of votes, reset index and print the result
avg_votes=goldenshows.groupby('imdb_score')['imdb_votes'].mean().reset_index()
print(avg_votes)


   imdb_score     imdb_votes
0         4.0    5277.583333
1         5.0    3143.942568
2         6.0    3481.717322
3         7.0    8727.068211
4         8.0   30299.460967
5         9.0  126904.109515


Now for the final step! Round the column with the averages, rename both columns, and print the dataframe in descending order.

In [24]:
# round column with averages
avg_votes['imdb_votes']=avg_votes['imdb_votes'].round(0)


# rename columns
avg_votes.rename(columns={'imdb_score':'avg_rating','imdb_votes':'avg_votes'},inplace=True)

# print dataframe in descending order
print(avg_votes.sort_values(by=['avg_votes'],ascending=False))

   avg_rating  avg_votes
5         9.0   126904.0
4         8.0    30299.0
3         7.0     8727.0
0         4.0     5278.0
2         6.0     3482.0
1         5.0     3144.0


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Excellent - you've calculated the mean values and sorted them the way it is easy to see the ones that got the highest number of votes..

The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Overall conclusion is an important part, where we should include the summary of the outcomes of the project.
    
Please note that you will need to write one for every project you make.